File: packages/graph-builder/src/types/tool.ts
/**
 * TypeScript types for LLM tool definitions.
 * These types mirror the backend models in graphflow_core/models/tool.py
 */

/**
 * Visibility mode for a property mapping.
 * - 'llm': Property is exposed to the LLM as a tool parameter
 * - 'runtime': Property is hidden from LLM, provided by memory/constant
 */
export type PropertyVisibility = 'llm' | 'runtime';

/**
 * Defines how a step property is mapped when used as a tool.
 */
export interface ToolPropertyMapping {
  /** Property key from the source step's config schema */
  source_property: string;

  /** Who controls this property */
  visibility: PropertyVisibility;

  /** If visibility='runtime': constant value or memory binding like {memory.api_key} */
  runtime_value?: string;

  /** If visibility='llm': parameter name exposed to LLM (defaults to source_property) */
  llm_parameter_name?: string;

  /** If visibility='llm': description shown to LLM for this parameter */
  llm_description?: string;

  /** If visibility='llm': JSON schema for this parameter (type, enum, etc.) */
  llm_schema?: Record<string, any>;

  /** If visibility='llm': whether this parameter is required */
  required?: boolean;
}

/**
 * Defines a tool that wraps an existing step for LLM use.
 */
export interface ToolDefinition {
  /** Unique identifier for this tool */
  id: string;

  /** Tool name visible to LLM (e.g., 'search_web', 'get_user') */
  name: string;

  /** Tool description for LLM explaining when/how to use it */
  description: string;

  /** Step type to wrap (e.g., 'http.HTTPGetStep', 'db_query') */
  source_step_type: string;

  /** How each step property is handled */
  property_mappings: ToolPropertyMapping[];

  /** Which output from the step to return to LLM */
  output_key?: string;

  /** Optional Python expression to transform output */
  output_transform?: string;
}

/**
 * A tool entry that wraps a step with inline definition.
 * Stored directly in the LLM step's tools array.
 */
export interface MappedStepTool {
  type: 'mapped_step';
  definition: ToolDefinition;
}

/**
 * A tool entry that uses a direct function definition (OpenAI format).
 * Allows users to define tools manually without mapping to steps.
 */
export interface FunctionTool {
  type: 'function';
  /** OpenAI-style function definition */
  function: {
    name: string;
    description: string;
    parameters: {
      type: 'object';
      properties: Record<string, any>;
      required?: string[];
    };
  };
}

/**
 * MCP Server configuration for connecting to an MCP server.
 */
export interface MCPServerConfig {
  /** Transport type: 'stdio', 'sse', or 'streamable_http' */
  transport: 'stdio' | 'sse' | 'streamable_http';

  /** Command to run for stdio transport */
  command?: string;

  /** Arguments for stdio command */
  args?: string[];

  /** Environment variables for stdio process */
  env?: Record<string, string>;

  /** Server URL for SSE or Streamable HTTP transport */
  url?: string;

  /** HTTP headers for SSE/Streamable HTTP */
  headers?: Record<string, string>;

  /** Connection/call timeout in seconds */
  timeout?: number;
}

/**
 * Definition of an MCP tool to expose to an LLM.
 */
export interface MCPToolDefinition {
  /** Unique identifier for this tool */
  id: string;

  /** Tool name visible to LLM (can differ from MCP tool name) */
  name: string;

  /** Tool description for LLM */
  description: string;

  /** Original tool name on the MCP server */
  mcp_tool_name: string;

  /** How each MCP tool parameter is handled (LLM vs runtime) */
  property_mappings: ToolPropertyMapping[];

  /** Key for the output in the result */
  output_key?: string;
}

/**
 * A tool entry that wraps an MCP server tool.
 */
export interface MCPTool {
  type: 'mcp';
  server: MCPServerConfig;
  definition: MCPToolDefinition;
}

/**
 * Union type for all tool entry types in an LLM step's tools array.
 */
export type ToolEntry = MappedStepTool | FunctionTool | MCPTool;

/**
 * Helper to check if a tool entry is a mapped step tool.
 */
export function isMappedStepTool(entry: ToolEntry): entry is MappedStepTool {
  return entry.type === 'mapped_step';
}

/**
 * Helper to check if a tool entry is a function tool.
 */
export function isFunctionTool(entry: ToolEntry): entry is FunctionTool {
  return entry.type === 'function';
}

/**
 * Helper to check if a tool entry is an MCP tool.
 */
export function isMCPTool(entry: ToolEntry): entry is MCPTool {
  return entry.type === 'mcp';
}

/**
 * Get LLM-controlled parameters from a tool definition.
 */
export function getLLMParameters(tool: ToolDefinition): ToolPropertyMapping[] {
  return tool.property_mappings.filter(m => m.visibility === 'llm');
}

/**
 * Get runtime-provided parameters from a tool definition.
 */
export function getRuntimeParameters(tool: ToolDefinition): ToolPropertyMapping[] {
  return tool.property_mappings.filter(m => m.visibility === 'runtime');
}

/**
 * Convert a tool definition to OpenAI function schema format.
 */
export function toOpenAIFunctionSchema(tool: ToolDefinition): Record<string, any> {
  const properties: Record<string, any> = {};
  const required: string[] = [];

  for (const mapping of getLLMParameters(tool)) {
    const paramName = mapping.llm_parameter_name || mapping.source_property;

    // Build parameter schema
    const paramSchema = mapping.llm_schema || { type: 'string' };
    if (mapping.llm_description) {
      paramSchema.description = mapping.llm_description;
    }

    properties[paramName] = paramSchema;

    if (mapping.required !== false) {
      required.push(paramName);
    }
  }

  return {
    type: 'function',
    function: {
      name: tool.name,
      description: tool.description,
      parameters: {
        type: 'object',
        properties,
        required,
      },
    },
  };
}

/**
 * Create a new empty tool definition with defaults.
 */
export function createEmptyToolDefinition(id?: string): ToolDefinition {
  return {
    id: id || `tool_${Date.now()}`,
    name: '',
    description: '',
    source_step_type: '',
    property_mappings: [],
    output_key: 'result',
  };
}

/**
 * Create a mapped step tool from a definition.
 */
export function createMappedStepTool(definition: ToolDefinition): MappedStepTool {
  return {
    type: 'mapped_step',
    definition,
  };
}

/**
 * Create an MCP tool from server config and definition.
 */
export function createMCPTool(server: MCPServerConfig, definition: MCPToolDefinition): MCPTool {
  return {
    type: 'mcp',
    server,
    definition,
  };
}

/**
 * Create an empty MCP server config with defaults.
 */
export function createEmptyMCPServerConfig(): MCPServerConfig {
  return {
    transport: 'stdio',
    timeout: 30,
  };
}

/**
 * Create an empty MCP tool definition with defaults.
 */
export function createEmptyMCPToolDefinition(id?: string, mcpToolName?: string): MCPToolDefinition {
  return {
    id: id || `mcp_tool_${Date.now()}`,
    name: mcpToolName || '',
    description: '',
    mcp_tool_name: mcpToolName || '',
    property_mappings: [],
    output_key: 'result',
  };
}

File: packages/graph-core/graphflow_core/models/tool.py
"""Tool definition models for LLM tool support."""

from typing import Literal, Optional, List, Dict, Any
from pydantic import BaseModel, Field


class ToolPropertyMapping(BaseModel):
    """
    Defines how a step property is mapped when used as a tool.

    Each property can be either:
    - LLM-controlled: The LLM provides this value as a tool parameter
    - Runtime-provided: The value comes from memory or a constant
    """

    source_property: str = Field(
        ...,
        description="Property key from the source step's config schema"
    )

    visibility: Literal["llm", "runtime"] = Field(
        ...,
        description="Who controls this property: 'llm' for tool parameter, 'runtime' for hidden"
    )

    # For runtime properties
    runtime_value: Optional[str] = Field(
        None,
        description="If visibility='runtime': constant value or memory binding like {memory.api_key}"
    )

    # For LLM properties
    llm_parameter_name: Optional[str] = Field(
        None,
        description="If visibility='llm': parameter name exposed to LLM (defaults to source_property)"
    )

    llm_description: Optional[str] = Field(
        None,
        description="If visibility='llm': description shown to LLM for this parameter"
    )

    llm_schema: Optional[Dict[str, Any]] = Field(
        None,
        description="If visibility='llm': JSON schema for this parameter (type, enum, etc.)"
    )

    required: bool = Field(
        True,
        description="If visibility='llm': whether this parameter is required"
    )


class ToolDefinition(BaseModel):
    """
    Defines a tool that wraps an existing step for LLM use.

    Tools are created by:
    1. Selecting a source step type (e.g., http_get, db_query)
    2. Configuring which properties the LLM controls vs runtime provides
    3. Defining tool metadata (name, description) for the LLM

    The tool definition is stored in the graph JSON within the LLM step's config.
    At compile time, it's converted to framework-specific tool code.
    """

    id: str = Field(
        ...,
        description="Unique identifier for this tool within the step"
    )

    name: str = Field(
        ...,
        description="Tool name visible to LLM (e.g., 'search_web', 'get_user')"
    )

    description: str = Field(
        ...,
        description="Tool description for LLM explaining when/how to use it"
    )

    source_step_type: str = Field(
        ...,
        description="Step type to wrap (e.g., 'http.HTTPGetStep', 'db_query')"
    )

    property_mappings: List[ToolPropertyMapping] = Field(
        default_factory=list,
        description="How each step property is handled"
    )

    # Output configuration
    output_key: str = Field(
        "result",
        description="Which output from the step to return to LLM"
    )

    output_transform: Optional[str] = Field(
        None,
        description="Optional Python expression to transform output (e.g., 'result[\"data\"]')"
    )

    def get_llm_parameters(self) -> List[ToolPropertyMapping]:
        """Get all LLM-controlled parameters."""
        return [m for m in self.property_mappings if m.visibility == "llm"]

    def get_runtime_parameters(self) -> List[ToolPropertyMapping]:
        """Get all runtime-provided parameters."""
        return [m for m in self.property_mappings if m.visibility == "runtime"]

    def to_openai_function_schema(self) -> Dict[str, Any]:
        """
        Convert to OpenAI function calling format.

        Returns:
            Dict in OpenAI function schema format
        """
        properties = {}
        required = []

        for mapping in self.get_llm_parameters():
            param_name = mapping.llm_parameter_name or mapping.source_property

            # Build parameter schema
            param_schema = mapping.llm_schema or {"type": "string"}
            if mapping.llm_description:
                param_schema["description"] = mapping.llm_description

            properties[param_name] = param_schema

            if mapping.required:
                required.append(param_name)

        return {
            "type": "function",
            "function": {
                "name": self.name,
                "description": self.description,
                "parameters": {
                    "type": "object",
                    "properties": properties,
                    "required": required
                }
            }
        }


class MappedStepTool(BaseModel):
    """
    A tool entry in the LLM step's tools array that wraps a step.

    This is the format stored in the graph JSON:
    {
        "type": "mapped_step",
        "definition": { ... ToolDefinition ... }
    }
    """

    type: Literal["mapped_step"] = "mapped_step"
    definition: ToolDefinition


class FunctionTool(BaseModel):
    """
    A tool entry that uses a direct function definition (OpenAI format).

    This allows users to define tools manually without mapping to steps:
    {
        "type": "function",
        "function": { ... OpenAI function schema ... }
    }
    """

    type: Literal["function"] = "function"
    function: Dict[str, Any] = Field(
        ...,
        description="OpenAI-style function definition"
    )


# =============================================================================
# MCP (Model Context Protocol) Tool Models
# =============================================================================


class MCPServerConfig(BaseModel):
    """
    Configuration for connecting to an MCP server.

    Supports three transport types:
    - stdio: Spawn a local process (command + args)
    - sse: Server-Sent Events over HTTP
    - streamable_http: Streamable HTTP transport
    """

    transport: Literal["stdio", "sse", "streamable_http"] = Field(
        ...,
        description="Transport type for MCP connection"
    )

    # Stdio transport configuration
    command: Optional[str] = Field(
        None,
        description="Command to run for stdio transport (e.g., 'uvx', 'npx', 'python')"
    )
    args: Optional[List[str]] = Field(
        None,
        description="Arguments for stdio command"
    )
    env: Optional[Dict[str, str]] = Field(
        None,
        description="Environment variables for stdio process"
    )

    # HTTP transport configuration (SSE and Streamable HTTP)
    url: Optional[str] = Field(
        None,
        description="Server URL for SSE or Streamable HTTP transport"
    )
    headers: Optional[Dict[str, str]] = Field(
        None,
        description="HTTP headers for SSE/Streamable HTTP"
    )

    # Common configuration
    timeout: float = Field(
        30.0,
        description="Connection/call timeout in seconds"
    )

    def get_server_key(self) -> str:
        """
        Generate a unique key for this server configuration.
        Used to deduplicate connections to the same server.
        """
        if self.transport == "stdio":
            args_str = "_".join(self.args or [])
            return f"stdio_{self.command}_{args_str}"
        else:
            return f"{self.transport}_{self.url}"


class MCPToolDefinition(BaseModel):
    """
    Definition of an MCP tool to expose to an LLM.

    Similar to ToolDefinition but sources from an MCP server tool
    rather than a GraphFlow step.
    """

    id: str = Field(
        ...,
        description="Unique identifier for this tool within the step"
    )

    name: str = Field(
        ...,
        description="Tool name visible to LLM (can be different from MCP tool name)"
    )

    description: str = Field(
        "",
        description="Tool description for LLM (overrides MCP description if provided)"
    )

    mcp_tool_name: str = Field(
        ...,
        description="Original tool name on the MCP server"
    )

    property_mappings: List[ToolPropertyMapping] = Field(
        default_factory=list,
        description="How each MCP tool parameter is handled (LLM vs runtime)"
    )

    output_key: str = Field(
        "result",
        description="Key for the output in the result"
    )

    def get_llm_parameters(self) -> List[ToolPropertyMapping]:
        """Get all LLM-controlled parameters."""
        return [m for m in self.property_mappings if m.visibility == "llm"]

    def get_runtime_parameters(self) -> List[ToolPropertyMapping]:
        """Get all runtime-provided parameters."""
        return [m for m in self.property_mappings if m.visibility == "runtime"]

    def to_openai_function_schema(self) -> Dict[str, Any]:
        """
        Convert to OpenAI function calling format.

        Returns:
            Dict in OpenAI function schema format
        """
        properties = {}
        required = []

        for mapping in self.get_llm_parameters():
            param_name = mapping.llm_parameter_name or mapping.source_property

            # Build parameter schema
            param_schema = mapping.llm_schema or {"type": "string"}
            if mapping.llm_description:
                param_schema["description"] = mapping.llm_description

            properties[param_name] = param_schema

            if mapping.required:
                required.append(param_name)

        return {
            "type": "function",
            "function": {
                "name": self.name,
                "description": self.description,
                "parameters": {
                    "type": "object",
                    "properties": properties,
                    "required": required
                }
            }
        }


class MCPTool(BaseModel):
    """
    A tool entry in the LLM step's tools array that wraps an MCP server tool.

    This is the format stored in the graph JSON:
    {
        "type": "mcp",
        "server": { ... MCPServerConfig ... },
        "definition": { ... MCPToolDefinition ... }
    }

    Multiple MCPTool entries can reference the same server but different tools.
    At runtime, connections to the same server are shared.
    """

    type: Literal["mcp"] = "mcp"
    server: MCPServerConfig
    definition: MCPToolDefinition
